SmartDeviceLink/private/SDLProtocolReceivedMessageProcessor.m

```diff
@@ -0,0 +1,289 @@
+                        break;
+                }
+            }
+            break;
+            
+        case CONTROL_FRAME_INFO_STATE: {
+                SDLFrameInfo controlFrameInfo = 0x00;
+                // 8 bits for frame information
+                controlFrameInfo = currentByte;
+                self.state = SESSION_ID_STATE;
+                [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+                
+                // Check for errors. For these two frame types, the frame info should be 0x00
+                if (((self.frameType == SDLFrameTypeFirst) || (self.frameType == SDLFrameTypeSingle)) && (controlFrameInfo != 0x00)){
+                    self.state = ERROR_STATE;
+                    SDLLogD(@"Message frameType is out of spec");
+                }
+            }
+            break;
+            
+        case SESSION_ID_STATE:
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            self.state = DATA_SIZE_1_STATE;
+            break;
+        
+        // 32 bits for data size
+        case DATA_SIZE_1_STATE:
+            self.dataLength = 0;
+            self.dataLength += (UInt32)(currentByte & 0xFF) << 24;
+            self.state = DATA_SIZE_2_STATE;
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            break;
+            
+        case DATA_SIZE_2_STATE:
+            self.dataLength += (UInt32)(currentByte & 0xFF) << 16;
+            self.state = DATA_SIZE_3_STATE;
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            break;
+            
+        case DATA_SIZE_3_STATE:
+            self.dataLength += (UInt32)(currentByte & 0xFF) << 8;
+            self.state = DATA_SIZE_4_STATE;
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            break;
+            
+        case DATA_SIZE_4_STATE:
+            self.dataLength += (UInt32)(currentByte & 0xFF) << 0;
+            self.state = MESSAGE_1_STATE;
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Set the counter for the data pump.
+            self.dataBytesRemaining = self.dataLength;
+            
+            // Version 1 does not have a message ID so we skip to the data pump or the end.
+            if (self.version == 1) {
+                if (self.dataLength == 0) {
+                    [self resetState];
+                } else {
+                    self.state = DATA_PUMP_STATE;
+                }
+            }
+            
+            Byte headerSize = 0;
+            if (self.version == 1) {
+                headerSize = 8;
+            } else {
+                headerSize = 12;
+            }
+            
+            UInt32 maxMtuSize = 0;
+            if (self.version <= 2) {
+                maxMtuSize = 1500;
+            } else {
+                maxMtuSize = 131084;
```

Comment:
This is not always the case. To retrieve MTU size, you should call `[SDLGlobals sharedGlobals] mtuSizeForServiceType:]` because it can be dynamic in protocol version 5.

Reply:
> It relies on a dictionary of sizes for service type that is not set until after we pass the message to the router.

That is not correct. The size is negotiated in the initial connection phase, which uses a version 1 protocol packet. Did you see a failure in testing?

> However I should make a change for version 5
> I should add a small switch based of service type, that determines if we check mtu size, or if it is negotiated (which means we trust and use what the frame tells us)

That defeats the purpose of the state machine.

Reply:
After a lot of time studying the protocol the answer is that for version 5 we should not do this check.  The maxMtuSize for a v5 messages is either default, or given in the payload as part of the negotiation process.  This check is being done before we have received the payload, so we cannot know what the maxMtuSize should be.

per spec:
Control service - No maxMtuSize, so this check is meaningless.
RPC Service - maxMtuSize given in payload, so cannot do check before payload.
Audio Service - If maxMtuSize is given in payload, use that, else use default.  So cannot do check before payload.
Video Service - Same as Audio.
SDLServiceTypeBulkData - Protocol does not specify.  Assume same as Audio/Video due to similarities in type of data.

I have added an if statement to fix this.

Reply:
This is still not correct. You are misunderstanding what payload you should be looking at. I understand that you researched the spec, but you have misunderstood it. Please implement and test and I have suggested unless you have, in testing, found an error. Talk to me if you have more questions.

Reply:
I am implementing it now.  Apologies, I did not see the "That is not correct." comment due to page refresh.

Reply:
Implemented.  Seems to work.  I missed that initial connection is done with version 1, thus `SDLGlobals` can know a version and an `maxMtuSize`